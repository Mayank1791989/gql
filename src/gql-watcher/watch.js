/* @flow */
import globby from 'globby';
import _debounce from 'lodash/debounce';
import { type FileMatchConfig } from 'gql-config/types';
import parseFileMatchConfig from 'gql-config/parseFileMatchConfig';
import { type IWatcher, type WatchFile } from './types';
import boolValue from 'gql-shared/boolValue';

export type WatchOptions = {|
  rootPath: string,
  files: FileMatchConfig,
  Watcher: Class<IWatcher>,
  watch: boolean,
  onChange: (changedFiles: Array<WatchFile>) => void,
|};

export default function watch(options: WatchOptions) {
  const { glob, ignored } = parseFileMatchConfig(options.files);

  let watcher = null;
  const onReadyPromise = globby(glob, { cwd: options.rootPath })
    .then(matches => {
      const watchFiles = matches.map(file => ({ name: file, exists: true }));
      if (!boolValue(options.watch, true)) {
        return watchFiles;
      }
      // if in watchmode then wait for watcher to start
      return new Promise(resolve => {
        watcher = setupWatcher({
          rootPath: options.rootPath,
          onChange: options.onChange,
          Watcher: options.Watcher,
          glob,
          ignored,
        });
        watcher.on('ready', resolve);
      }).then(() => watchFiles);
    })
    .then(watchFiles => {
      options.onChange(watchFiles);
    })
    .catch(err => {
      throw err;
    });

  return {
    onReady(): Promise<void> {
      return onReadyPromise;
    },

    close(): void {
      if (watcher) {
        watcher.close();
      }
    },
  };
}

class WatchEventsBatcher {
  _listener: (files: Array<WatchFile>) => void;
  _queue = [];
  _activeTimeout = null;
  _waitTimeForBatching = 200 /* msec */;

  constructor(listener: (files: Array<WatchFile>) => void) {
    this._listener = listener;
  }

  __dispatchQueue = _debounce(() => {
    const events = this._queue;
    this._queue = [];

    if (events.length === 0) {
      return;
    }

    this._listener(events);
  }, this._waitTimeForBatching);

  add(eventOrEvents: WatchFile | Array<WatchFile>) {
    this._queue.push(
      ...(Array.isArray(eventOrEvents) ? eventOrEvents : [eventOrEvents]),
    );
    this.__dispatchQueue();
  }
}

function setupWatcher({ rootPath, onChange, glob, ignored, Watcher }) {
  const watchEventsBatcher = new WatchEventsBatcher(onChange);

  const watcher = new Watcher(rootPath, {
    glob,
    ignored,
  });

  watcher.on('change', filePath => {
    watchEventsBatcher.add({ name: filePath, exists: true });
  });

  watcher.on('add', filePath => {
    watchEventsBatcher.add({ name: filePath, exists: true });
  });

  watcher.on('delete', filePath => {
    watchEventsBatcher.add({ name: filePath, exists: false });
  });

  return watcher;
}
