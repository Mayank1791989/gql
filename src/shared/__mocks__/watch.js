/* @flow */
import glob from 'glob';

export default function watch({ rootPath, files, onChange }: any) {
  watch.__files = glob.sync(files, { cwd: rootPath }).map(_name => ({ name: _name, exists: true }));
  watch.__onChangeCallback = onChange;
}

watch.__triggerChange = (files) => {
  watch.__onChangeCallback(files || watch.__files);
};
