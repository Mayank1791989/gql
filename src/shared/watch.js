/* @flow */
import watchman from 'fb-watchman';
import { type WatchFile } from './types';
import { type FileMatchConfig, type Globs } from '../config/GQLConfig';

const globToMatchExpr = (glob: Globs): Array<any> => {
  if (typeof glob === 'string') {
    return ['match', glob, 'wholename'];
  }

  if (glob.length === 1) {
    return ['match', glob[0], 'wholename'];
  }

  // array
  return ['anyof', ...glob.map((g) => ['match', g, 'wholename'])];
};

export function toMatchExpression(match: FileMatchConfig) {
  // GlobPattern | Array<GlobPattern>
  if (typeof match === 'string' || Array.isArray(match)) {
    return globToMatchExpr(match);
  }

  // object form { match, exclude }
  const matchExpr = globToMatchExpr(match.include);
  const ignoreMatchExpr = match.ignore ? globToMatchExpr(match.ignore) : null;

  if (ignoreMatchExpr) {
    return ['allof', matchExpr, ['not', ignoreMatchExpr]];
  }

  return matchExpr;
}

type WatchOptions = {
  rootPath: string,
  files: FileMatchConfig,
  name: string,
  onChange: (changedFiles: Array<WatchFile>) => void,
};

export default function watch(options: WatchOptions) {
  const { rootPath, name, files, onChange } = options;
  // console.log(`Launching watch server for ${rootPath}`);
  const client = new watchman.Client();
  client.capabilityCheck(
    { optional: [], required: ['relative_root'] },
    (error) => {
      if (error) {
        console.error(error);
        client.end();
        return;
      }

      client.command(
        ['watch-project', rootPath],
        (err, { watch: _watch, relative_path: relativePath, warning }) => {
          if (err) {
            console.error('Error initiating watch:', err);
            return;
          }

          if (warning) {
            console.log('warning: ', warning);
          }

          const sub = {
            expression: toMatchExpression(files),
            fields: ['name', 'exists', 'type'],
            relative_root: relativePath, // eslint-disable-line camelcase
          };

          // console.log(`Watch established on ${_watch} ${files}`);

          const gqlFilesWatchSubscription = name;
          client.command(
            ['subscribe', _watch, gqlFilesWatchSubscription, sub],
            (subscribeError, resp) => {
              if (subscribeError) {
                console.error('failed to subscribe: ', subscribeError);
                return;
              }

              // $FlowDisableNextLine
              console.log(
                `[Watch established (${resp.subscribe})] \n\tbasePath: ${_watch} \n\tRelativePath: ${relativePath} \n\tFiles: ${JSON.stringify(files, 2, 2)}`,
              );
            },
          );

          client.on('subscription', (resp) => {
            if (resp.subscription !== gqlFilesWatchSubscription) {
              return;
            }
            onChange(resp.files);
          });
        },
      );
    },
  );
  return client;
}
