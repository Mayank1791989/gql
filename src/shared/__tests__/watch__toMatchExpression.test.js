/* @flow */
import { toMatchExpression } from '../watch';

test('string', () => {
  expect(toMatchExpression('files/**/*.gql')).toMatchSnapshot();
});

test('Array<string>', () => {
  expect(toMatchExpression(['files/**/*.gql', 'files2/**/*.gql'])).toMatchSnapshot();
});

test('Object<include: string>', () => {
  expect(toMatchExpression({
    include: 'files/**/*.gql',
  })).toMatchSnapshot();
});

test('Object<include: Array<string>>: single', () => {
  expect(toMatchExpression({
    include: ['files/**/*.gql'],
  })).toMatchSnapshot();
});

test('Object<match: Array<string>>', () => {
  expect(toMatchExpression({
    include: ['files/**/*.gql', 'files2/**/*gql'],
  })).toMatchSnapshot();
});

test('Object<include: Array<string>, ignore: string>', () => {
  expect(toMatchExpression({
    include: ['files/**/*.gql', 'files2/**/*gql'],
    ignore: '**/test/**/*.gql',
  })).toMatchSnapshot();
});

test('Object<match: Array<string>, ignore: Array<string>>', () => {
  expect(toMatchExpression({
    include: ['files/**/*.gql', 'files2/**/*gql'],
    ignore: ['**/test/**/*.gql'],
  })).toMatchSnapshot();
});
