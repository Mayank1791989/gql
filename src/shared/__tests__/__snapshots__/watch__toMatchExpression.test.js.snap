exports[`test Array<string> 1`] = `
Array [
  "anyof",
  Array [
    "match",
    "files/**/*.gql",
    "wholename",
  ],
  Array [
    "match",
    "files2/**/*.gql",
    "wholename",
  ],
]
`;

exports[`test Object<include: Array<string>, ignore: string> 1`] = `
Array [
  "allof",
  Array [
    "anyof",
    Array [
      "match",
      "files/**/*.gql",
      "wholename",
    ],
    Array [
      "match",
      "files2/**/*gql",
      "wholename",
    ],
  ],
  Array [
    "not",
    Array [
      "match",
      "**/test/**/*.gql",
      "wholename",
    ],
  ],
]
`;

exports[`test Object<include: Array<string>>: single 1`] = `
Array [
  "match",
  "files/**/*.gql",
  "wholename",
]
`;

exports[`test Object<include: string> 1`] = `
Array [
  "match",
  "files/**/*.gql",
  "wholename",
]
`;

exports[`test Object<match: Array<string>, ignore: Array<string>> 1`] = `
Array [
  "allof",
  Array [
    "anyof",
    Array [
      "match",
      "files/**/*.gql",
      "wholename",
    ],
    Array [
      "match",
      "files2/**/*gql",
      "wholename",
    ],
  ],
  Array [
    "not",
    Array [
      "match",
      "**/test/**/*.gql",
      "wholename",
    ],
  ],
]
`;

exports[`test Object<match: Array<string>> 1`] = `
Array [
  "anyof",
  Array [
    "match",
    "files/**/*.gql",
    "wholename",
  ],
  Array [
    "match",
    "files2/**/*gql",
    "wholename",
  ],
]
`;

exports[`test string 1`] = `
Array [
  "match",
  "files/**/*.gql",
  "wholename",
]
`;
